api/devicerepository.proto

```diff
@@ -298,11 +298,24 @@ message ListEndDeviceModelsRequest {
 }
 
 message GetTemplateRequest {
+  // Identifiers to uniquely identify a LoRaWAN end device profile.
+  message EndDeviceProfileIdentifiers {
+    // VendorID managed by the LoRa Alliance, as defined in TR005.
+    uint32 vendor_id = 1 [(validate.rules).uint32 = {gte: 1}];
+
+    // ID of the LoRaWAN end device profile. The naming of this field follows the TR005 specification.
+    uint32 vendor_profile_id = 2;
+    }
+
   // Application identifiers.
   ApplicationIdentifiers application_ids = 1 [(validate.rules).message.required = true];
 
   // End device version information.
+  // EndDeviceProfileIdentifiers take precedence over this field if both are specified.
   EndDeviceVersionIdentifiers version_ids = 2;
+
+  // End device profile identifiers.
+  EndDeviceProfileIdentifiers end_device_profile_ids = 3;
```

Comment:
You could also move these two into a oneof; proto-wise that's a backwards compatible change.

Reply:
The `oneof` won't be compatible with the gRPC gateway I think.

Reply:
Indeed. I didn't think that it was backwards compatible. I'd very much prefer to use a oneof.

Reply:
@adriansmares can you give an example of this breaking compatibility? Because I can't think of any.

The only thing that changes here is the generated code. On the wire (both as protobuf and as JSON) setting `version_ids` (or not setting) looks exactly the same, regardless of whether it's in a oneof or not.

Reply:
Oneof fields cannot be used in the paths for the HTTP paths of the gRPC gateway, or at least so it seems. We've had the same discussion for the storage integration, where you could get data for the whole application, or for a single end device (so a `oneof` made perfect sense from a protobuf perspective).

Take the following diff that introduces the `oneof`:
```diff
diff --git a/api/devicerepository.proto b/api/devicerepository.proto
index f203d27d3..18c64b12d 100644
--- a/api/devicerepository.proto
+++ b/api/devicerepository.proto
@@ -305,17 +305,19 @@ message GetTemplateRequest {
 
     // ID of the LoRaWAN end device profile. The naming of this field follows the TR005 specification.
     uint32 vendor_profile_id = 2;
-    }
+  }
 
   // Application identifiers.
   ApplicationIdentifiers application_ids = 1 [(validate.rules).message.required = true];
 
-  // End device version information.
-  // EndDeviceProfileIdentifiers take precedence over this field if both are specified.
-  EndDeviceVersionIdentifiers version_ids = 2;
+  oneof ids {
+    // End device version information.
+    // EndDeviceProfileIdentifiers take precedence over this field if both are specified.
+    EndDeviceVersionIdentifiers version_ids = 2;
 
-  // End device profile identifiers.
-  EndDeviceProfileIdentifiers end_device_profile_ids = 3;
+    // End device profile identifiers.
+    EndDeviceProfileIdentifiers end_device_profile_ids = 3;
+  }
 }
 
 message GetPayloadFormatterRequest {
```

It works exactly as you expect, and the PBs compile happily. Then at runtime we try to actually use one of the HTTP paths:

```
adriansmares@Adrian-Stefans-MacBook-Pro lorawan-stack % curl 'http://localhost:1885/api/v3/dr/applications/dev1/brands/adeunis/models/analog/1.03/EU_863_870/template' -H 'Authorization: Bearer MFRWG.FW64YQR7SU5UG52S7IGILBES675MXJGYHJMVKCQ.XJXQQBPARCOS7JLICFKBSOEDUGODLUJ5COJOAHSJOL5C2AI675OA'
{"code":3,"message":"type mismatch, parameter: version_ids.model_id, error: field already set for ids oneof"}%
```

Reply:
> Oneof fields cannot be used in the paths for the HTTP paths of the gRPC gateway, or at least so it seems. 

We also use oneof fields for the the collaborator in the IS, and that has always worked just fine:

https://github.com/TheThingsNetwork/lorawan-stack/blob/fd41a7df32f47c0cf3bf4af6b5f65e4ea3a239d0/api/application_services.proto#L36-L45